### DISCUSSION TITLE
Crashing in Firefox when dealing with large conversations and long generations

### DISCUSSION THREAD

**Discussion by ThomasBaruzier** [danny-avila/LibreChat#4417] [Troubleshooting]
### What happened?

Firefox tab is crashing around 50% of the time when dealing with long generations over long context queries

### Steps to Reproduce

Use LibreChat
Paste 30-50k token codebase, ask questions, wait. I measured 45k tokens in my current conversation, generations included.
Works better with DeepSeek API, since it's slow and the crash is random, the likelihood of crashes is higher. Odds skyrockets when generating in the background and actively using another tab. Finally, there is a 20-30% crash when the generation ends.

### What browsers are you seeing the problem on?

Firefox

### Relevant log output

No logs, Firefox does not seem to log tab crashes

Nothing meaningful in /app/api/logs/error-2024-10-15.log either:
```shell
{"level":"error","message":"[handleAbortError] AI response error; aborting request: [GoogleGenerativeAI Error]: Error fetching from https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro-002:streamGenerateContent?alt=sse: [429 Too Many Requests] Resource has been exhausted (e.g. check quota).","stack":"Error: [429 Too Many Requests] Resource has been exhausted (e.g. check quota).\n    at makeRequest (/app/api/node_modules/@google/generative-ai/dist/index.js:263:19)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async generateContentStream (/app/api/node_modules/@google/generative-ai/dist/index.js:649:22)\n    at async GoogleClient.getCompletion (/app/api/app/clients/GoogleClient.js:699:22)\n    at async GoogleClient.sendCompletion (/app/api/app/clients/GoogleClient.js:881:13)\n    at async GoogleClient.sendMessage (/app/api/app/clients/BaseClient.js:574:24)\n    at async AskController (/app/api/server/controllers/AskController.js:130:20)\n    at async /app/api/server/routes/ask/google.js:23:5"}
{"level":"error","message":"[handleAbortError] AI response error; aborting request: [GoogleGenerativeAI Error]: Error fetching from https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-pro-exp-0827:streamGenerateContent?alt=sse: [500 Internal Server Error] An internal error has occurred. Please retry or report in https://developers.generativeai.google/guide/troubleshooting","stack":"Error: [500 Internal Server Error] An internal error has occurred. Please retry or report in https://developers.generativeai.google/guide/troubleshooting\n    at makeRequest (/app/api/node_modules/@google/generative-ai/dist/index.js:263:19)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async generateContentStream (/app/api/node_modules/@google/generative-ai/dist/index.js:649:22)\n    at async GoogleClient.getCompletion (/app/api/app/clients/GoogleClient.js:699:22)\n    at async GoogleClient.sendCompletion (/app/api/app/clients/GoogleClient.js:881:13)\n    at async GoogleClient.sendMessage (/app/api/app/clients/BaseClient.js:574:24)\n    at async AskController (/app/api/server/controllers/AskController.js:130:20)\n    at async /app/api/server/routes/ask/google.js:23:5"}
```

I am sorry this bug report does not offer proper logging, I'd be glad to help with some guidance.

### Screenshots

![image](https://github.com/user-attachments/assets/91d75874-08a8-49a6-99e0-03cabd73f891)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
I’m not able to reproduce it on Firefox, and Chrome doesn’t seem to have this issue.

**Reply by ThomasBaruzier** [reply]
![image](https://github.com/user-attachments/assets/ae21b6c9-e0f8-419f-989c-fe07409de917)

Maybe this could be an issue?

**Reply by danny-avila** [reply]
Can you try the same amount of text on chrome? Firefox is known for this issue and even when it doesn’t crash I can observe the difference in speed. There may be even OS-specific fixes you can try.

**Reply by ThomasBaruzier** [reply]
![image](https://github.com/user-attachments/assets/932c53d2-9d4c-435b-8efc-4979a77fbaa8)

Ram is a bit better, but CPU and FPS suffer more in the latest stable chromium. No crash for now, but it's inconvenient to have my laptop fans blow at 60-70% each time, while scrolling at 5-10 FPS.
I have a i5-12450H and a 4060 Max-Q for reference

**Reply by ThomasBaruzier** [reply]
![image](https://github.com/user-attachments/assets/3754a739-2824-4629-ab65-281a1c3d9086)

For one 8k response, I got 18 minutes of CPU time

**Reply by danny-avila** [reply]
Can you measure the performance on chatgpt or claude.ai?

**Reply by ThomasBaruzier** [reply]
Yep, i'll try later today

**Reply by ThomasBaruzier** [reply]
![image](https://github.com/user-attachments/assets/257f287a-b782-444b-a9fe-6fb629c84d28)

Since I can't do long context on ChatGPT or Claude web interfaces as a free user, I tried the same setup on chat.deepseek.com.
As you can see, there is a 40% CPU and 250-300 MB RAM usage. Plus, the 18-minute CPU screenshot I sent was a 8k DeepSeek completion, so this comparison should be fair. The website can generate only 4k tokens, so I ran it twice. At the end, I got 4 minutes of CPU time, minus 30s, because that's what I started with before generating.

![image](https://github.com/user-attachments/assets/8c0ed5ab-1e4d-4eb9-95ec-050487abf9d7)

I don't mind the RAM usage of LibreChat, what bothers me is CPU usage while generating, because CPU usage outside this scope stays low (unless for search, but that's ok too). Maybe the Markdown code blocks are being regenerated/re-rendered on each new token?

**Reply by RAPHCVR** [reply]
@danny-avila I have the same issue, with deepseek r1 on firefox or edge, it's very chatty, it very often never finish the chat because the tab crash before. On deepseek chat app, same length but no issue at all.

**Reply by danny-avila** [reply]
@RAPHCVR I noticed a performance issue when using Openrouter specifically, not sure if it applies to this thread (since it was happening with few messages) but it's fixed as of this PR: https://github.com/danny-avila/LibreChat/pull/5456

**Reply by RAPHCVR** [reply]
@danny-avila Unfortunately, I’m still experiencing the issue even after the latest update. My browser crashes whenever DeepSeek generates a lot of text. I’m not the only one facing this—several friends who have deployed their own instances of LibreChat along with DeepSeek are encountering the same problem. Were you able to test this on your end?
![image](https://github.com/user-attachments/assets/fe2b73ba-b820-428f-95e5-303df041d1df)

**Reply by RAPHCVR** [reply]
On edge it seems it's a bit better, but the ram usage is still very high (800MB to 2GB during generation). I also mentioned there is styling in Edge/Chromium there is not in firefox ("Thoughts" pannels for example).

**Reply by ThomasBaruzier** [reply]
@RAPHCVR 
I used chromium as a workaround :(

**Reply by danny-avila** [reply]
I’m not sure why it would be worse on Firefox vs. chrome, but I have no RAM issues on either and I’ve never experienced a crash when generating a lot of text.

**Reply by ThomasBaruzier** [reply]
Today the amount of crashes is low enough for me, but I was getting a really high amount of them (75-90% chance when using DeepSeek v2.5 to go up to 8k output). Today I do the same, but with a 2-5% chance of crashing.
The difference between now and before is that I don't work anymore with a program that filled up my ram up to 10 of my 16gbs on my Linux laptop.
So either the performance has been improved or the available ram is a key factor.

**Reply by RAPHCVR** [reply]
I have a pretty powerfull pc but still have this issue. It is indeed much less present on chromium. Also, as I said before, there is some styling that are present on chromium but not on firefox (the "thought" widget for example, does not appear on firefox it seems). And finally, the issue is also present with friends of mine using deepseek R1 too on firefox.

**Reply by thomas-333** [reply]
I'm seeing issues as well with Firefox. Hard to pin down exactly, but during long generations at times I can hear my CPU fans spin up and the other day while using Deepseek R1 all my 32GB of memory was fully utilised. I know this doesn't help debugging, but if I can isolate it I'll post again.

**Reply by danny-avila** [reply]
Adding to the conversation my notes.

After doing some profiling of render speeds, it seems react-markdown will just slow down on re-render with large texts: https://github.com/orgs/remarkjs/discussions/1027

This is not new to me but I under-estimated the amount of text where performance is felt.

On re-render, the outlier is the Markdown component in terms of performance (increases render time by amount of text)
![image](https://github.com/user-attachments/assets/944dc1f7-f498-4bc7-b97d-a3ba9097d0f3)

Normally, AI do not go on streaming a lot of content in one Markdown block unless specifically prompted or has innate propensity to do so (like `deepseek-reasoner`).

This issue presents itself less when using [Agents](https://librechat.ai/docs/features/agents) as they use a new message paradigm of content blocks.

We can do the same for legacy endpoints to break up into blocks, separated by length by the rough amount of text it takes to cover a regular-sized screen.

The only challenge that remains is making sure the new "reasoning" UI is still accounted for when broken up into blocks.
